### PR TITLE
Add extractors for Status of HTTP Response

### PR DESCRIPTION
**Pull request by spockz**
> When writing response classifiers, and other code that deals with responses,
> in a generic manner, it is currently inconvenient to match against specific
> ranges of errors such as Server Errors or Client errors.
> 
> Solution
> 
> By adding extractors for the different categories (Unknown, Informational,
> Successful, Redirection, Client Error, and Server Error) it becomes possible
> to easily determine in which category a response falls.
> 
> Result
> 
> Determining whether a response was succesful is as easy as matching
> on Successful(_).

It can be debated whether we should also provide an extractor that works on the `Response` directly, and if so, where the code should be placed.

### CODE REVIEW

File: CHANGES [hunk 1]
```diff
@@ -7,8 +7,16 @@ Note that ``RB_ID=#`` correspond to associated messages in commits.
 6.x
 -----
 
+6.34.0
```
Comment (kevinoliver) [inline CHANGES]: leave this next version out please. our release tooling automatically bumps the 6.x to the next version when its release time.

File: finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala [hunk 2]
```diff
@@ -29,7 +29,47 @@ case class Status(code: Int) {
...
+    def unapply(status: Status): Option[Status] =
+      Some(status).filter(s => s.code < 100 || s.code >= 600)
+  }
+
+  object Informational {
+    def unapply(status: Status): Option[Status] =
+      inRange(100, 200)(status)
+  }
+
+  object Successful{
```
Comment (kevinoliver) [inline finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala]: nit: space before the {

File: finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala [hunk 3]
```diff
@@ -29,7 +29,47 @@ case class Status(code: Int) {
...
+  private[finagle] def inRange(lower: Int, upper: Int)
+                              (status: Status): Option[Status] =
+    Some(status).filter(s => s.code >= lower && s.code < upper)
+
+  object UnknownStatus {
+    def unapply(status: Status): Option[Status] =
+      Some(status).filter(s => s.code < 100 || s.code >= 600)
+  }
+
+  object Informational {
```
Comment (kevinoliver) [inline finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala]: mind adding trivial scaladoc to these public objects? can be as simple as:

```
/** Matches when the status code is between 100–199. */
```

**Comment by kevinoliver** [comment]
this will be a nice addition. thanks.

**Comment by mosesn** [comment]
hey, would you mind adding a commit where the first line of the commit message is the title of this RB, prefixed with `finagle-http: `?

File: finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala [hunk 4]
```diff
@@ -29,7 +29,47 @@ case class Status(code: Int) {
...
 
+
+
+
+
 object Status {
+
+  @inline
+  private[finagle] def inRange(lower: Int, upper: Int)
+                              (status: Status): Option[Status] =
```
Comment (taylorleese) [inline finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala]: Do we really need this to be curried?

File: finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala [hunk 5]
```diff
@@ -29,7 +29,47 @@ case class Status(code: Int) {
...
+    Some(status).filter(s => s.code >= lower && s.code < upper)
```
Comment (taylorleese) [inline finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala]: On the off chance somebody passes in a null Status it would be nice if this were ```Option(status)``` here rather than ```Some(status)```. Same comment applies to line 45.

**Comment by spockz** [comment]
@mosesn: Added the commit per request

File: finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala (hunk 2 above)
Comment (spockz) [inline finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala]: fixed

File: finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala (hunk 3 above)
Comment (spockz) [inline finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala]: Done

File: CHANGES (hunk 1 above)
Comment (spockz) [inline CHANGES]: Done

File: finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala (hunk 4 above)
Comment (spockz) [inline finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala]: No, not really. It was like that because I wanted to write the unapply function like this:
```scala
def unapply: Status => Option[Status] =
  inRange(100,200)
```
But it appears that leads to several compile errors.

File: finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala (hunk 5 above)
Comment (spockz) [inline finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala]: Good point, fixed.

File: finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala [hunk 6]
```diff
@@ -1,20 +1,20 @@
...
- * - it simplifies construction (users only supply the code)
- * - it avoids the need to validate user-defined reason phrases
- * - it omits the possibility of statuses with duplicate reason phrases
- *
- * The only downside is that we lose the ability to create custom statuses with
- * "vanity" reason phrases, but this should be tolerable.
- *
- * For Java-friendly enums, see [[com.twitter.finagle.http.Statuses]].
- */
+  * Represents an HTTP status code.
```
Comment (kevinoliver) [inline finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala]: sorry to nitpick: but our style used for scaladoc is 1 less indent on *s.

```
/**
 * ...
 * ...
 */
```

**Comment by kevinoliver** [comment]
lgtm @spockz once you cleanup the scaladoc style.

**Comment by spockz** [comment]
Should we/I add an extra extractor to match on a `Response` as well? Maybe in a different PR?

File: finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala (hunk 6 above)
Comment (spockz) [inline finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala]: Is there something that automatically formats this way? My auto-layout destroys it back to the layout in the PR. Was this also described somewhere?

Comment (mosesn) [inline finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala]: Yep, we adhere to the comment style here: https://twitter.github.io/effectivescala/#Formatting-Comments

File: finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala [hunk 7]
```diff
@@ -29,7 +29,65 @@ case class Status(code: Int) {
...
+  /**
+   * Matches when the status code is between 500 and 600.
+   */
+  object ServerError {
+    def unapply(status: Status): Option[Status] =
+      inRange(500, 600, status)
+  }
+
+  @inline
+  private[finagle] def inRange(lower: Int, upper: Int,
```
Comment (mosesn) [inline finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala]: can you just put all of this on one line? we typically don't wrap until 100 columns.

File: finagle-http/src/test/scala/com/twitter/finagle/http/StatusTest.scala [hunk 8]
```diff
@@ -15,4 +16,45 @@ class StatusTest extends FunSuite {
     assert(Status(599).reason == "Server Error")
     assert(Status(601).reason == "Unknown Status")
   }
+
+  test("Pattern match usability check") {
```
Comment (mosesn) [inline finagle-http/src/test/scala/com/twitter/finagle/http/StatusTest.scala]: can you break this up into separate tests for each thing you're testing?  ie, "matches unknown" "matches informational" etc.

File: finagle-http/src/test/scala/com/twitter/finagle/http/StatusTest.scala [hunk 9]
```diff
@@ -15,4 +16,45 @@ class StatusTest extends FunSuite {
...
+    null match {
```
Comment (mosesn) [inline finagle-http/src/test/scala/com/twitter/finagle/http/StatusTest.scala]: why test this?  null isn't a valid input, is it?  if it is, we should instead accept an Option[Status], imo.

File: CHANGES [hunk 10]
```diff
@@ -7,8 +7,13 @@ Note that ``RB_ID=#`` correspond to associated messages in commits.
 6.x
 -----
 
+New Features
+~~~~~~~~~~~~
+  * finagle-http: Introduce convience extractors to pattern match `c.t.f.http.Response.status`
```
Comment (mosesn) [inline CHANGES]: s/convience/convenience

**Comment by mosesn** [comment]
LGTM except for a few nits.

File: finagle-http/src/test/scala/com/twitter/finagle/http/StatusTest.scala (hunk 9 above)
Comment (spockz) [inline finagle-http/src/test/scala/com/twitter/finagle/http/StatusTest.scala]: `null` isn't a status, therefore, it shouldn't be matched by any of the category matchers.

Comment (mosesn) [inline finagle-http/src/test/scala/com/twitter/finagle/http/StatusTest.scala]: Right, but my impression is that in scala we typically assume that null is not a valid argument to pass.  If an instance might be null, it's up to the user to use it safely.  If that weren't the case, we would need to null-check every argument ever.

We could make the argument that there's no harm in adding a test for null, but since folks often look at tests as example code, I'd be worried that this will end up being cargo-culted.

File: finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala [hunk 11]
```diff
@@ -29,7 +29,54 @@ case class Status(code: Int) {
...
+  /** Matches when the status code is between 500 and 600. */
+  object ServerError {
+    def unapply(status: Status): Option[Status] =
+      inRange(500, 600, status)
+  }
+
+  @inline
+  private[finagle] def inRange(lower: Int, upper: Int,
+                               status: Status): Option[Status] =
+    Option(status).filter(s => s.code >= lower && s.code < upper)
```
Comment (mosesn) [inline finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala]: this should be `Some(status)`, I think.

Comment (spockz) [inline finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala]: Fine by me. :)

Comment (taylorleese) [inline finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala]: I had commented that it should be ```Option(status)``` in case it was null by some chance. That's what I usually do stylistically out of paranoia, but not a huge deal either way.

File: finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala [hunk 12]
```diff
@@ -29,7 +29,54 @@ case class Status(code: Int) {
...
+
 object Status {
+
+  /**
+   * Matches when the status code isn't in range of any of the  categories: {{Informational}},
+   * {{Successful}}, {{Redirection}}, {{ClientError}}, {{ServerError}}.
+   */
+  object UnknownStatus {
+    def unapply(status: Status): Option[Status] =
+      Option(status).filter(s => s.code < 100 || s.code >= 600)
```
Comment (taylorleese) [inline finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala]: This should match line 77 to be consistent. I prefer ```Option```, but I may be the only one so consistency is more important.

File: finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala [hunk 13]
```diff
@@ -29,7 +29,54 @@ case class Status(code: Int) {
...
 }
 
+
 object Status {
+
+  /**
+   * Matches when the status code isn't in range of any of the  categories: {{Informational}},
+   * {{Successful}}, {{Redirection}}, {{ClientError}}, {{ServerError}}.
+   */
+  object UnknownStatus {
```
Comment (vkostyukov) [inline finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala]: How about to rename it just `Unknown`? We already have prefix `Status` so `Status.UnknownStats` feels redundant.

File: finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala (hunk 11 above)
Comment (mosesn) [inline finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala]: Ah, we typically only use `Option` in finagle when we're using an API that can definitely return `null` sometimes.  We treat `Option` like the Nullable annotation in findbugs–if it's not marked Option, it definitely exists.

File: finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala (hunk 13 above)
Comment (spockz) [inline finagle-http/src/main/scala/com/twitter/finagle/http/Status.scala]: Good point, fixed.

**Comment by mosesn** [comment]
LGTM

**Comment by vkostyukov** [comment]
I'm pulling this internally. Will close this PR once it's merged.

**Comment by vkostyukov** [comment]
This has been merged internally. Thanks @spockz!
